### ISSUE TITLE
Reveal hidden amp-list after button click with layout container does not work

### ISSUE THREAD

**Issue by dritter** [ampproject/amphtml#23769]
Hi there!

We would like to reveal an amp-list after a button was clicked. Unfortunately this does not work. The request to the backend is never made. If I remove the `changeToLayoutContainer()` (Button "Just Show List" in example) this works. I also tried to bind hidden and set a state without success.

Here is an example from the playground:

```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>amp-list</title>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <!-- ## Setup -->
  <!-- Import the `amp-list` component ... -->
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <!-- ... and the `amp-mustache` component in the header. -->
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <!-- Import the `amp-bind` component for dynamically changing the content of an `amp-list`. -->
  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
  <link rel="canonical" href="https://amp.dev/documentation/examples/components/amp-list/index.html">
  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  
</head>
<body>
  <button on="tap:test-list.changeToLayoutContainer(), test-list.show">
    Show List with Layout Container
  </button>
  
  <button on="tap:test-list.show">
    Just Show List
  </button>
  
  <amp-list id="test-list" hidden layout="responsive" height="100" width="100"
      src="/static/samples/json/examples.json"
      binding="no">
    <template type="amp-mustache">
      <div><a href="{{url}}">{{title}}</a></div>
    </template>
  </amp-list>

</body>
</html>
```

**Comment by dritter** [comment]
A workaround for that problem would be to wrap the list and make the wrapper invisible. Making that wrapper show/hide still does not work.

```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>amp-list</title>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <!-- ## Setup -->
  <!-- Import the `amp-list` component ... -->
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <!-- ... and the `amp-mustache` component in the header. -->
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <!-- Import the `amp-bind` component for dynamically changing the content of an `amp-list`. -->
  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
  <link rel="canonical" href="https://amp.dev/documentation/examples/components/amp-list/index.html">
  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <style amp-custom>
    .hide-me {
    max-height: 0;
    opacity: 0;
    overflow: hidden;
    transition: max-height .15s ease,opacity .15s ease;
    visibility: hidden;
    width: 100%;
}
  </style>
</head>
<body>
  <amp-state id="mystate">
    <script type="application/json">
      {
      	"hidden": true
      }
    </script>
  </amp-state>
  
  <button on="tap:test-list.changeToLayoutContainer(), AMP.setState({mystate: {hidden: false}})">
    Show List with Layout Container
  </button>
  
  <button on="tap:AMP.setState({mystate: {hidden: false}})">
    Just Show List
  </button>
  
  <div class="hide-me"
       [class]="mystate.hidden ? 'hide-me' : ''">
    <amp-list id="test-list"
              layout="responsive"
              height="100"
              width="100"
              src="/static/samples/json/examples.json"
              binding="no">
      <template type="amp-mustache">
        <div><a href="{{url}}">{{title}}</a></div>
      </template>
    </amp-list>
  </div>

</body>
</html>
```

**Comment by dritter** [comment]
That workaround is not flawless as well: Fetching the list endpoint is only triggered on scroll. If you have a huge container in between the button and the list, the list won't show up on click (and won't even if you scroll down afterwards):

```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>amp-list</title>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <!-- ## Setup -->
  <!-- Import the `amp-list` component ... -->
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <!-- ... and the `amp-mustache` component in the header. -->
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <!-- Import the `amp-bind` component for dynamically changing the content of an `amp-list`. -->
  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
  <link rel="canonical" href="https://amp.dev/documentation/examples/components/amp-list/index.html">
  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <style amp-custom>
    .hide-me {
    max-height: 0;
    opacity: 0;
    overflow: hidden;
    transition: max-height .15s ease,opacity .15s ease;
    visibility: hidden;
    width: 100%;
}
  </style>
</head>
<body>
  <amp-state id="mystate">
    <script type="application/json">
      {
      	"hidden": true
      }
    </script>
  </amp-state>
  
  <button on="tap:test-list.changeToLayoutContainer(), AMP.setState({mystate: {hidden: false}})">
    Show List with Layout Container
  </button>
  
  <button on="tap:AMP.setState({mystate: {hidden: false}})">
    Just Show List
  </button>
  
  <div style="height: 10000px; border: 1px solid red">
    If you click on the button before scrolling
    the list into the viewport, the request is
    never made. Even if you scroll down afterwards.
    <br />
    <br />
    To make this work, you need to scroll down to 
    the list first, and click the button afterwards.
    <br />
    <br />
    Have an eye on your network tab in the dev tools,
    to see when the request is triggered.
  </div>
  
  <div class="hide-me"
       [class]="mystate.hidden ? 'hide-me' : ''">
    <amp-list id="test-list"
              layout="responsive"
              height="100"
              width="100"
              src="/static/samples/json/examples.json"
              binding="no">
      <template type="amp-mustache">
        <div><a href="{{url}}">{{title}}</a></div>
      </template>
    </amp-list>
  </div>

</body>
</html>
```

**Comment by cathyxz** [comment]
I was going to take a closer look at this, but for now:
 
The reason why your initial code didn't work is because `<amp-list>` shouldn't be changed to layout container prior to load by design. Technically, it doesn't work because layout container sizes to its contents and a hidden `<amp-list>` has size 0x0. Elements with dimensions 0x0 will not have their layout callback called, which for `<amp-list>` means that it won't fetch data. 

One of our goals on AMP is to prevent content shifting (esp without user intent) and this is why `<amp-list>` could not support layout `CONTAINER` in the first place. Basically, if you're able to look at content below the `<amp-list>` and `<amp-list>` pushes content up or down due to a height change after load, that creates a jarring user experience. 

Basically, it was intentional to not enable layout container before the initial fetch to prevent content shifting (i.e. a 0x0 element resizing and pushing content down). Your code here presents an interesting use case. Technically, since the list was hidden prior, user interaction was required for changing to layout container, therefore you can perhaps argue that there was clear user intent behind the fetch, which makes it ok to resize. 

Question: If you add a placeholder element with a fixed size, does that solve the problem?

**Comment by dritter** [comment]
Thanks for the insights. That was very interesting and helpful.

> you can perhaps argue that there was clear user intent behind the fetch, which makes it ok to resize.

Yes, I do think the user intent is clear here. The user clicks a button and IMHO showing an element that was hidden before is in the interest of the user..

> Question: If you add a placeholder element with a fixed size, does that solve the problem?

Awesome! That indeed fixes the problem. I added `<div placeholder>Loading ...</div>` to my initial example (even without a fixed size) and it worked. For a better DX it would be nice to have the same behaviour even without the placeholder though.
